app/src/main/java/com/keylesspalace/tusky/fragment/NotificationsFragment.java

```diff
@@ -545,6 +565,161 @@ public void onNotificationContentCollapsedChange(boolean isCollapsed, int positi
+                    break;
+                case FOLLOW:
+                    itemId = R.id.filter_follows;
+                    break;
+                case REBLOG:
+                    itemId = R.id.filter_boosts;
+                    break;
+                case MENTION:
+                    itemId = R.id.filter_mentions;
+                    break;
+            }
+            if (itemId != View.NO_ID) {
+                MenuItem item = menu.getMenu().findItem(itemId);
+                if (item != null)
+                    item.setChecked(false);
+            }
+        }
+        menu.setOnMenuItemClickListener(item -> {
+            if (item.isCheckable()) {
+                item.setChecked(!item.isChecked());
+
+                //Prevent popup menu close on item click
+                item.setShowAsAction(MenuItem.SHOW_AS_ACTION_COLLAPSE_ACTION_VIEW);
+                item.setActionView(anchor);
+                item.setOnActionExpandListener(new MenuItem.OnActionExpandListener() {
+                    @Override
+                    public boolean onMenuItemActionExpand(MenuItem item) {
+                        return false;
+                    }
+
+                    @Override
+                    public boolean onMenuItemActionCollapse(MenuItem item) {
+                        return false;
+                    }
+                });
+                return false;
+            } else {
+                //Apply filter changes and close the menu
+                applyFilterChanges(menu.getMenu());
+                return true;
+            }
+        });
+        menu.show();
+    }
+
+    private void applyFilterChanges(Menu menu) {
+        boolean isChanged = updateNotificationFilter(menu, R.id.filter_favourites, Notification.Type.FAVOURITE);
+        isChanged |= updateNotificationFilter(menu, R.id.filter_follows, Notification.Type.FOLLOW);
+        isChanged |= updateNotificationFilter(menu, R.id.filter_boosts, Notification.Type.REBLOG);
+        isChanged |= updateNotificationFilter(menu, R.id.filter_mentions, Notification.Type.MENTION);
+        if (isChanged) {
+            saveNotificationsFilter();
+            fullyRefresh();
+        }
+
+    }
+
+    private Gson getNotificationsGson() {
+        return new GsonBuilder()
+                .registerTypeAdapter(Notification.Type.class, new Notification.NotificationTypeAdapter())
+                .create();
+    }
+
+    private void loadNotificationsFilter() {
+        String json = preferences.getString("notifications.filter", "[]");
+        Set<Notification.Type> set = getNotificationsGson().fromJson(json, new TypeToken<Set<Notification.Type>>() {
+        }.getType());
+        if (set != null)
+            notificationFilter.addAll(set);
+    }
+
+    private void saveNotificationsFilter() {
+        String json = getNotificationsGson().toJson(notificationFilter, new TypeToken<Set<Notification.Type>>() {
+        }.getType());
```

Comment:
please de/serialize the settings in a more efficient way, e.g. by concatenating the active ones to a string seperated by some special character. Gson is way too expensive for this simple task.